tests/unit/src/util/Git.ts

```diff
@@ -0,0 +1,350 @@
+let existsSyncStub: SinonStub;
+let chmodSyncStub: SinonStub;
+
+registerSuite({
+	name: 'util/Git',
+
+	before() {
+		promiseSpawnStub = stub();
+		promiseExecStub = stub();
+		execStub = stub();
+		loggerStub = stub();
+		toStringStub = stub();
+		existsSyncStub = stub();
+		chmodSyncStub = stub();
+	},
+
+	after() {
+		cleanupModuleMocks();
+	},
+
+	beforeEach() {
+		Module = loadModule('src/util/Git', {
+			'./process': {
+				promiseSpawn: promiseSpawnStub,
+				promiseExec: promiseExecStub,
+				exec: execStub
+			},
+			'../log': {
+				logger: {
+					info: loggerStub
+				}
+			},
+			'./streams': {
+				toString: toStringStub
+			},
+			fs: {
+				existsSync: existsSyncStub,
+				chmodSync: chmodSyncStub
+			}
+		});
+
+		git = new Module();
+	},
+
+	afterEach() {
+		promiseSpawnStub.reset();
+		promiseExecStub.reset();
+		execStub.reset();
+		loggerStub.reset();
+		toStringStub.reset();
+		existsSyncStub.reset();
+		chmodSyncStub.reset();
+	},
+
+	'constructor': {
+		'with params'() {
+			const gitWithArgs = new Module('dir', 'file');
+			assert.equal(gitWithArgs.cloneDirectory, 'dir');
+			assert.equal(gitWithArgs.keyFile, 'file');
+		},
+		'default params'() {
+			assert.equal(git.cloneDirectory, process.cwd());
+			assert.equal(git.keyFile, env.keyFile());
+		}
+	},
+
+	async add() {
+		promiseExecStub.withArgs('git add file1 file2', {
+			silent: false,
+			cwd: git.cloneDirectory
+		}).returns('pass');
+
+		const actual = git.add('file1', 'file2');
+
+		assert.instanceOf(actual, Promise);
+		assert.strictEqual(await actual, 'pass');
+	},
+
+	async assert() {
+		git.isInitialized = () => false;
+		git.assert('url').then(() => assert.fail(), (error: Error) => {
+			assert.strictEqual(error.message,
+				`Repository is not initialized at "${ git.cloneDirectory }"`);
+		});
+
+		git.isInitialized = () => true;
+		git.getConfig = stub().withArgs('remote.origin.url')
+			.returns('url');
+
+		git.assert('other_url').then(() => assert.fail(), (error: Error) => {
+			assert.strictEqual(error.message,
+				'Repository mismatch. Expected "url" to be "other_url".');
+		});
+		assert.doesNotThrow(async () => await git.assert('url'));
+	},
+
+	async checkout() {
+		promiseExecStub.withArgs('git checkout 1.2.3', {
+			silent: false,
+			cwd: git.cloneDirectory
+		}).returns(Promise.resolve('pass'));
+
+		const actual = git.checkout('1.2.3');
+
+		assert.instanceOf(actual, Promise);
+		assert.strictEqual(await actual, 'pass');
+	},
+
+	clone: {
+		'If clone directory is not set; throws'() {
+			delete git.cloneDirectory;
+			git.clone('url').then(() => assert.fail(), (error: Error) => {
+				assert.equal(error.message,
+					'A clone directory must be set');
+			});
+		},
+
+		// logger should be called once internally and once within execSSHAgent
+		async 'Not initialized; logger called 1x, git.url === url'() {
+			const url = 'url';
+			git.isInitialized = () => false;
+			await git.clone(url);
+			assert.isTrue(loggerStub.calledTwice);
```

Comment:
Let's not test the logger. It's a brittle test as a change to logging in any method will break this test.

Instead, let's test that assert is not called

```
const assertSpy = spy(git, 'assert');
assert.false(assertSpy.called);
```

We should also test that execSSHAgent is called once